tools/data-api/internal/repositories/services.go

```diff
@@ -29,8 +32,7 @@ type ServicesRepositoryImpl struct {
 
 func NewServicesRepository(directory string, serviceType ServiceType, serviceNames *[]string) ServicesRepository {
```

Comment:
> Given there's going to be multiple sub-directories of API Definitions within the ./api-definitions directory - it'd be worth supporting that sooner rather than later.

The services repository directory is constructed out of the root dir for the api definitions and a service type via which we route by so as far as I'm concerned this is already supported?

> Presumably we'll want to identify any sub-directories containing a metadata.json file in this method, so that we can validate that each of the Services are unique

Two points here:

1. Given there's already a mechanism for determining which service type to read the api definitions in for I don't currently see a value in identifying sub-dirs that contain a `metadata.json`, I guess at most to validate that the `Source` value in the `metadata.json` corresponds to the api definitions we're wanting to load.
2. Checking that there aren't multiple definitions for the same service makes sense

I think 2. this is a good isolated small task that can and in the interest of time be completed in a separate PR. I've left a TODO in the method.

Reply:
For 1:

The relationship between `ServiceType` and `Source` within the `metadata.json` isn't (necessarily) 1:1 - we're going to need to port over [the existing hand-defined data](https://github.com/hashicorp/pandora/tree/main/data/Pandora.Definitions.HandDefined/AADB2C) for Resource Manager, which (since it can't be regenerated) will want to be sourced from a different path.

I was working on the assumption that we'd discover any Sources of API Definitions within this path, which means we don't need to change the Data API to serve a new set of data (e.g. if we needed to handcraft something for MSGraph) - rather than having the directories hard-coded (which feels like it'd be hard to discover, since the `directory` argument here is the path to `./api-definitions`)?

I guess we could do the same thing by parsing a configuration file within `./api-definitions` to define which directory houses what type of data - but having the Data API know both the path to the `api-definitions` folder (provided), and then the sub-directories within it as a part of the Data API feels more magical to me?

Reply:
It would help if you could provide a complete example of the directory structure you're expecting in `api-definitions` because now I'm unsure we have the same understanding.

Reply:
Sure, so I'm working on the assumption that `./api-definitions` looks like this:

```
 $ tree
.
├── README.md
├── handwritten-resource-manager
│   ├── AADB2C
│   │   ├── 2020-01-01
│   │   │   ├── ApiVersionDefinition.json
│   │   │   └── Example
│   │   │       ├── Constant-Example.json
│   │   │       ├── Model-ExampleResource.json
│   │   │       ├── Operation-Get.json
│   │   │       └── ResourceId-Example.json
│   │   └── ServiceDefinition.json
│   └── metadata.json
├── microsoft-graph
│   ├── Applications
│   │   ├── V1
│   │   │   ├── ApiVersionDefinition.json
│   │   │   └── Example
│   │   │       ├── Constant-Example.json
│   │   │       ├── Model-ExampleResource.json
│   │   │       ├── Operation-Get.json
│   │   │       └── ResourceId-Example.json
│   │   └── ServiceDefinition.json
│   └── metadata.json
└── resource-manager
    ├── Advisor
    │   ├── 2020-01-01
    │   │   ├── ApiVersionDefinition.json
    │   │   └── Example
    │   │       ├── Constant-Example.json
    │   │       ├── Model-ExampleResource.json
    │   │       ├── Operation-Get.json
    │   │       └── ResourceId-Example.json
    │   ├── 2022-10-01
    │   │   ├── ApiVersionDefinition.json
    │   │   └── Example
    │   │       ├── Constant-Example.json
    │   │       ├── Model-ExampleResource.json
    │   │       ├── Operation-Get.json
    │   │       └── ResourceId-Example.json
    │   └── ServiceDefinition.json
    └── metadata.json
```

That is, multiple directories can contain the API Definitions for Resource Manager:

* `./api-definitions/handwritten-resource-manager` - contains the handwritten API Definitions.
* `./api-definitions/resource-manager` - contains the generated API Definitions.

And (at present) only one directory for the API Definitions for Microsoft Graph:

* `./api-definitions/microsoft-graph` - contains the generated API Definitions.

(Aside: @manicminer wants to update how MS Graph outputs it's data, such that it'll be using a single `microsoft-graph` endpoint with versioning done ala ARM (e.g. `./api-definitions/microsoft-graph/Applications/(V1|Beta)` - but that's not something we need to handle now / can be dealt with in the future).

Each of the sub-directories contains a `metadata.json` file which we can use to determine which Service Type this is associated with:

```json
{
	"dataSource": "AzureResourceManager",
	"sourceInformation": "Azure/azure-rest-api-specs",
	"gitRevision": "13ef12f5de907f32bb6cac13123824c48c911411"
}
```

Which we can use to build up a canonical set of Data for each Service Type - so that we can serve this via a single endpoint.

It's worth noting that when building up the list of supported Services, we'll want to validate that each Service exists only in a single set of data (that is, `Storage` can exist in both `handwritten-resource-manager` and `resource-manager` - but not both). In the event we need to hand-define a Service for some reason, using a more specific name for the handwritten set of data shouldn't be a big deal.